DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGenerator.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Org.BouncyCastle.Asn1.X9;
+using Org.BouncyCastle.Crypto.Digests;
+using Org.BouncyCastle.Crypto.Generators;
+using Org.BouncyCastle.Crypto.Parameters;
+using Org.BouncyCastle.Math;
+
+namespace DIGNDB.App.SmitteStop.API.Services
+{
+    public class RollingKeyPairGenerator
+    {
+        private readonly byte[] _masterKeyBytes;
+
+        public X9ECParameters ECParameters { get; private set; }
+
+        public RollingKeyPairGenerator(byte[] masterKeyBytes, X9ECParameters ecParameters)
+        {
+            _masterKeyBytes = masterKeyBytes;
+            ECParameters = ecParameters;
+        }
+
+        public RollingKeyPairGenerator(X509Certificate2 certificate, X9ECParameters ecParameters)
+            : this(RetrievePrivateKeyBytes(certificate), ecParameters)
+        {
+        }
+
+        private static byte[] RetrievePrivateKeyBytes(X509Certificate2 certificate)
+        {
+            var ecdsaPrivateKey = certificate.GetECDsaPrivateKey();
+            if (ecdsaPrivateKey != null)
+            {
+                return ecdsaPrivateKey.ExportParameters(true).D;
+            }
+
+            var rsaPrivateKey = certificate.GetRSAPrivateKey();
+            if (rsaPrivateKey != null)
+            {
+                // TODO: find a better way to extract bytes for RSA key
+                return Encoding.UTF8.GetBytes(rsaPrivateKey.ToXmlString(true));
+            }
+
+            throw new NotSupportedException($"Unsupported private key for certificate {certificate.Thumbprint}");
+        }
+
+        public BigInteger GetPrivateKey(string keyId)
+        {
+            return GeneratePrivateKey(_masterKeyBytes, long.Parse(keyId), ECParameters);
+        }
+
+        public (BigInteger privateKey, ECPublicKeyParameters publicKey) GenerateKeyPairForInterval(long keyIntervalNumber)
+        {
+            var privateKey = GeneratePrivateKey(_masterKeyBytes, keyIntervalNumber, ECParameters);
+            var publicKey = CalculatePublicKey(privateKey, ECParameters);
+            return (privateKey, publicKey);
+        }
+
+        private static BigInteger GeneratePrivateKey(byte[] masterKeyBytes, long keyIntervalNumber, X9ECParameters ecParameters)
+        {
+            var keyByteSize = (int) Math.Ceiling(ecParameters.Curve.Order.BitCount / 8.0);
+            var counter = 0;
+
+            BigInteger privateKey;
+            do
+            {
+                var privateKeyBytes = GeneratePrivateKeyBytes(keyByteSize, masterKeyBytes, keyIntervalNumber, counter);
+                privateKey = new BigInteger(privateKeyBytes);
+                counter++;
```

Comment:
Add if clause to ensure this never can run infinitely, and throw an exception if above some high limit.  Just a sanity check.

Reply:
Should do the same in the verification solution. The probability of having to run the loop multiple times should be low, so I guess a cap of 100 should be enough, could go even higher just to be sure..